src/integrity-checker.js

```diff
@@ -0,0 +1,133 @@
+/* @flow */
+
+import type Config from './config.js';
+import * as constants from './constants.js';
+import {registryNames} from './registries/index.js';
+import Lockfile from './lockfile/wrapper.js';
+import * as crypto from './util/crypto.js';
+import * as fs from './util/fs.js';
+import {sortAlpha} from './util/misc.js';
+
+
+const invariant = require('invariant');
+const path = require('path');
+
+export type IntegrityCheckResult = {
+  integrityFileMissing: boolean,
+  integrityHashMatches?: boolean,
+  missingPatterns: Array<string>,
+};
+
+/**
+ *
+ */
+export default class InstallationIntegrityChecker {
+  constructor(config: Config) {
+    this.config = config;
+  }
+
+  config: Config;
+
+  /**
+   * Get the location of an existing integrity hash. If none exists then return the location where we should
+   * write a new one.
+   */
+
+  async _getIntegrityHashLocation(): Promise<string> {
+    // build up possible folders
+    const possibleFolders = [];
+    if (this.config.modulesFolder) {
+      possibleFolders.push(this.config.modulesFolder);
+    }
+
+    // ensure we only write to a registry folder that was used
+    for (const name of registryNames) {
+      const loc = path.join(this.config.cwd, this.config.registries[name].folder);
+      possibleFolders.push(loc);
+    }
+
+    // if we already have an integrity hash in one of these folders then use it's location otherwise use the
+    // first folder
+    const possibles = possibleFolders.map((folder): string => path.join(folder, constants.INTEGRITY_FILENAME));
+    let loc = possibles[0];
```

Comment:
> on a similar note but possibly out of scope for this PR, shouldn't the integrity file be different for each registry rather than global to all of them?

We use one type of registry per project, unlikely there will be a bower_modules and node_modules in one folder, so it should be fine to have only one integrity there.